lib/change_stream.js

```diff
@@ -13,27 +13,128 @@ const CHANGE_DOMAIN_TYPES = {
   DATABASE: Symbol('Database'),
   CLUSTER: Symbol('Cluster')
 };
+class ResumeTokenTracker extends EventEmitter {
+  constructor(changeStream, options) {
+    super();
+    this.changeStream = changeStream;
+    this.options = options;
+    this._postBatchResumeToken = undefined;
+  }
+
+  set resumeToken(token) {
+    this._resumeToken = token;
+    // NOTE: Event is for internal use only, and is not part of public API
+    this.emit('tokenChange');
+  }
+
+  get resumeToken() {
+    return this._resumeToken;
+  }
+
+  init() {
+    this._resumeToken = this.options.startAfter || this.options.resumeAfter;
+    this._operationTime = this.options.startAtOperationTime;
+    this._init = true;
+  }
+
+  resumeInfo() {
+    const resumeInfo = {};
+
+    if (this._init && this.resumeToken) {
+      resumeInfo.resumeAfter = this.resumeToken;
+    } else if (this._init && this._operationTime) {
+      resumeInfo.startAtOperationTime = this._operationTime;
+    } else {
+      if (this.options.startAfter) {
+        resumeInfo.startAfter = this.options.startAfter;
+      }
+
+      if (this.options.resumeAfter) {
+        resumeInfo.resumeAfter = this.options.resumeAfter;
+      }
+
+      if (this.options.startAtOperationTime) {
+        resumeInfo.startAtOperationTime = this.options.startAtOperationTime;
+      }
+    }
+
+    return resumeInfo;
+  }
+
+  onResponse(postBatchResumeToken, operationTime) {
+    if (this.changeStream.isClosed()) {
+      return;
+    }
+    const cursor = this.changeStream.cursor;
+    if (!postBatchResumeToken) {
+      if (
+        !(this.resumeToken || this._operationTime || cursor.bufferedCount()) &&
+        cursor.server &&
+        cursor.server.ismaster.maxWireVersion >= 7
+      ) {
+        this._operationTime = operationTime;
+      }
+    } else {
```

Comment:
Is it possible that `onResponse` could be called with `null` for both `postBatchResumeToken` and `operationTime`?

Reply:
I don't think it is possible, b/c any scenario where the operationTime is blank will error b/c that server does not support changeStreams.